### PR TITLE
Ukernels: simplify the architecture-specific bitcode build.

### LINKED ISSUE #16000: [LLVMCPU] Fix linking between modules with different target attributes
Just recording thoughts before holidays...

Around LLVMCPUTarget.cpp, when we load and link different bitcode modules, we are currently just removing some function attributes to make it more or less work, https://github.com/openxla/iree/blob/9cde4e3dc589ab783757b863a747477bc6fafc91/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/Builtins/UKernel.cpp#L35-L44 ,  but it's not very rigorous, and it needs to become more rigorous in order to overcome two issues we're having:
* When we remove all function target attributes including target CPU features, and the code inside a function was actually relying on a CPU feature (e.g. was using an intrinsic provided by that CPU feature), we are effectively introducing undefined behavior. It mostly just works, until this week it didn't --- one of the bugs I've dealt with this week was, in int4-quantized Llama2, some avx512vnni LLVM intrinsic was causing a `llvm_unreachable` to be reached in the LLVM x86 backend.
* What we are currently doing with the attributes-stripping isn't entirely successful at one of its main goals, which is to enable perfect inlining of ukernel code into dispatch functions. It's working on x86 but not on arm64. It's probably a matter of being more rigorous about how we bring attributes in good alignment between caller and callee.

I'm thinking that the solution is, rather than *dropping* attributes, we just need to *add* all the attributes from the `llvm::TargetMachine`. Then, by construction, all functions that are actually legal to call on that target machine will have exactly the same attribute (give or take any preexisting difference in the attributes that they had before). And functions that have additional attributes are not legal to call on this target machine anyway.

As a first approximation to make this concrete, I'm toying with a change like this:

```diff
-static void removeTargetAttributes(llvm::Module &module) {
+static void overrideTargetAttributes(llvm::Module &module, const llvm::TargetMachine& targetMachine) {
   for (auto &func : module.functions()) {
-    func.removeFnAttr("target-cpu");
-    func.removeFnAttr("tune-cpu");
-    func.removeFnAttr("target-features");
+    func.addFnAttr("target-cpu", targetMachine.getTargetCPU());
+    func.addFnAttr("target-features", targetMachine.getTargetFeatureString());
   }
 }
```

I'll get back to this in early 2024... I have been mulling something like this for months as I knew about the issue about ukernels not getting fully inlined on arm64, but now that I know that it's also causing UB manifesting as compiler crashes on Llama2 on x86, this became high priority.

FYI @benvanik @Max191

### PR DESCRIPTION
**Pull request by bjacob**
This PR basically re-does how we build and link architecture-specific and generic bitcode, for CPU ukernels.

The main change is that we now build one separate ukernel bitcode file for each CPU architecture, and it's the only thing that `iree-compile` needs to load and link to. Before, it had to load and link together 3 different bitcode modules, two of which were actually compiled as WebAssembly for genericity, and it involved weak symbols. Full inlining is really important for performance here, and in this system, it depended on two fragile mechanisms:
1. Reinterpreting a WebAssembly module as a native module.
2. Weak symbols.

That was done out of a concern for sharing generic code across same-bitness architectures, but that turned out be substantially more complex in both the ukernels and the compiler, and to suffer from problems described in #16000, which this PR fixes. Namely, the inlining across these modules and down to the calling dispatch function was brittle, never worked outside of x86, and to make even that work, we had to do some illegal target-attribute stripping that was effectively UB and caused issues described in https://github.com/llvm/llvm-project/issues/78206 (originally filed as a LLVM bug but then understood to be caused by our own illegal attribute-stripping).

So, we now build completely separate bitcode for each target CPU architecture that we are supporting in `iree-compile`. At the moment, that's roughly 20k of code for each additional architecture for which we don't have dedicated optimized code paths. For comparison, each architecture that does have dedicated code weighs roughly 100k. I found that surprisingly large, until I remembered that IR is far more verbose than object code. There are ways to control that if and when that becomes a major concern:
* We could further trim the supported CPU architectures, but note that we gain simplicity from not doing so at the moment.
* We could leave bitcode files separate rather than embedding them in `iree-compile`, allowing each user to obtain their own for their target architecture.
* We could just embed the ukernels *source* code, as pure C is so cheap to compile - the main cost would be now having to essentially build/ship `clang` as part of iree-compile, but supporting only C and not C++ shouldn't be that bad.
  * One problem that that would also solve is that at the moment, for CPU architectures for which `iree-compile` allows the user to pick from multiple ABIs (looking at you, RISC-V --- you need to grow up, pick an ABI and settle in life), compiling ukernels from source at that time would make it trivial to have ukernels built for the right ABI. For now in this PR we just pick whichever ABI our tests are currently using.

Note: this PR is broken into a few thematic commits to help reviewing.

Fixes #16000

### CODE REVIEW

**Comment by bjacob** [comment]
Debugging that dispatch code size regression.

**Comment by bjacob** [comment]
got it, needed AlwaysInline attr on ukernel symbols. actually now i remember @MaheshRavishankar had that in his original ukernel linking code. For some reason that wasn't needed with the WebAsm stuff but is needed now that we're going back to native arch.

EDIT - done, and as it seems that AlwaysInline implies the same DCE-ability of non-called functions as InternalLinkage provided, but unlike InternalLinkage can be applied upon loading the bitcode before linking it, this yields another massive simplification, as the tracking of ukernel symbols to apply InternalLinkage to post-linking accounted for much of the remaining code.

**Comment by bjacob** [comment]
Yesss! Now the updated benchmarks summary shows pure improvements on both latency and size.

@benvanik PTAL

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/Builtins/UKernel.cpp [hunk 1]
```diff
@@ -27,96 +27,39 @@ loadUKernelBitcodeFile(StringRef filename, llvm::LLVMContext &context) {
...
-}
-
 llvm::Expected<std::unique_ptr<llvm::Module>>
-loadUKernelArchEntryPointsBitcode(llvm::TargetMachine *targetMachine,
-                                  llvm::LLVMContext &context) {
+loadUKernelBitcode(llvm::TargetMachine *targetMachine,
+                   llvm::LLVMContext &context) {
   const char *archName =
       getIreeArchNameForTargetTriple(targetMachine->getTargetTriple());
   char filename[64];
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/Builtins/UKernel.cpp]: let's fix this to be C++ while in here - the arbitrary limit here will be tricky to debug if someone comes along with a longer arch string - `std::string("ukernel_bitcode_") + archName + ".bc"`

**Review by benvanik — APPROVED** [review]

**Comment by benvanik** [comment]
very cool! hooray for across-the-board wins - including developer productivity :)

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/Builtins/UKernel.cpp (hunk 1 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/Builtins/UKernel.cpp]: done

**Pull request merged**